src/commands/index.ts

```diff
@@ -0,0 +1,17 @@
+export const allCommands = new Map<string, Command>();
+
+/**  Install commands here:  **/
+import { help } from './help';
+_add(help);
+
+export function _add(cmd: Command): void {
+	const name = cmd.name;
+
+	if (allCommands.has(name)) {
+		throw new TypeError(
+			`Failed to add command '${name}' when a command with that name was already added`
+		);
+	}
+
+	allCommands.set(name, cmd);
+}
```

Comment:
Couldn't this index just automatically compile the list of commands? I mean, all I has to do is use
`fs.readdirSync('./').filter(file => file.endsWith('.ts') && !file.endsWith('.test.ts' && file != 'index.ts');`
And then we don't have to update this every time we add a new command.

Reply:
That's not how TS works.

Since the files are transpiled before runtime, file introspection for `.ts` or `.test.ts` or even `.d.ts` files in the commands directory will not work.

We \*could\* check for `.js` files, but we'd have no way of knowing at runtime whether those files are properly-formatted command definitions. To avoid having to handle potentially-catastrophic errors that may never happen, we import the file. This way, TypeScript has a chance to check that the commands we want are correctly structured, with the proper properties and `execute` method present. A direct import at the top of the file also informs code analyzers (and possible code bundlers, should we go that route) what files we use, so that those command files do not get stripped in a bundle or skipped in a security analysis as dead code.

Reply:
TypeScript also ensures that the command actually registered, so long as we update the index. This simplifies unit tests down the road. Filesystem introspection and dynamic imports are both slow and error-prone, so I'd want lots of tests around those, and constructing them would be difficult.